File: python/1774. Closest Dessert Cost/solution.py
class Solution:
    def closestCost(self, baseCosts: List[int], toppingCosts: List[int], target: int) -> int:
        self.closestCost = float(-inf)

        def helper(currentCost, baseCosts, toppingCosts, index, target):
            if abs(currentCost - target) == abs(self.closestCost - target):
                self.closestCost = min(self.closestCost, currentCost)
            elif abs(currentCost - target) < abs(self.closestCost - target):
                self.closestCost = currentCost

            if index == len(toppingCosts):
                return

            helper(currentCost, baseCosts, toppingCosts, index + 1, target)
            helper(currentCost + toppingCosts[index], baseCosts, toppingCosts, index + 1, target)
            helper(currentCost + 2 * toppingCosts[index], baseCosts, toppingCosts, index + 1, target)


        for baseCost in baseCosts:
            helper(baseCost, baseCosts, toppingCosts, 0, target)

        return self.closestCost

File: java/1774. Closest Dessert Cost/Solution.java
class Solution {
    private int[] baseCosts;
    private int[] toppingCosts;
    private int target;
    private int closestCost;

    public int closestCost(int[] baseCosts, int[] toppingCosts, int target) {
        this.baseCosts = baseCosts;
        this.toppingCosts = toppingCosts;
        this.target = target;
        this.closestCost = baseCosts[0];

        for (int baseCost : baseCosts)
            helper(baseCost, 0);
        return closestCost;

    }

    private void helper(int currentCost, int index) {
        if (Math.abs(currentCost - target) == Math.abs(closestCost - target))
            closestCost = Math.min(closestCost, currentCost);
        else if (Math.abs(currentCost - target) < Math.abs(closestCost - target))
            closestCost = currentCost;
        if (index == toppingCosts.length) return;
        helper(currentCost, index + 1);
        helper(currentCost + toppingCosts[index], index + 1);
        helper(currentCost + 2 * toppingCosts[index], index + 1);
    }
}